\contextbox{Publication}{
This chapter corresponds to the extended version of the following publication: \fullcite{DBLP:conf/wabi/GagieGM21}.
}
\todo[inline]{Move the appendix back to the main body. Add back acknowledgement. Add update on the research since the publication}

\newcommand{\inputXBWT}[1]{\input{Part_Two/xbwt/#1}}
\newcommand{\figXBWT}[2]{\includegraphics[width=#1\textwidth]{Part_Two/xbwt/#2}}

\begin{small}

Compressed full-text indexes are one of the main success stories of bioinformatics data structures but even they struggle to handle some DNA readsets.  This may seem surprising since, at least when dealing with short reads from the same individual, the readset will be highly repetitive and, thus, highly compressible.  If we are not careful, however, this advantage can be more than offset by two disadvantages: first, since most base pairs are included in at least tens reads each, the uncompressed readset is likely to be at least an order of magnitude larger than the individual's uncompressed genome; second, these indexes usually pay some space overhead for each string they store, and the total overhead can be substantial when dealing with millions of reads.


The most successful compressed full-text indexes for readsets so far are based on the Extended Burrows-Wheeler Transform (EBWT) and use a sorting heuristic to try to reduce the space overhead per read, but they still treat the reads as separate strings and thus may not take full advantage of the readset's structure.  For example, if we have already assembled an individual's genome from the readset, then we can usually use it to compress the readset well: e.g., we store the gap-coded list of reads' starting positions; we store the list of their lengths, which is often highly compressible; and we store information about the sequencing errors, which are rare with short reads.  There is nowhere, however, where we can plug an assembled genome into the EBWT.

In this paper we show how to use one or more assembled or partially assembled genome as the basis for a compressed full-text index of its readset.  Specifically, we build a labelled tree by taking the assembled genome as a trunk and grafting onto it the reads that align to it, at the starting positions of their alignments.  Next, we compute the eXtended Burrows-Wheeler Transform (XBWT) of the resulting labelled tree and build a compressed full-text index on that. Although this index can occasionally return false positives, it is usually much more compact than the alternatives.
Following the established practice for datasets with many repetitions, we compare different full-text indices by looking at the number of runs in the transformed strings. For a human Chr19 readset our preliminary experiments show that eliminating separators characters from the EBWT reduces the number of runs by 19\%, from 220 million to 178 million, and using the XBWT reduces it by a further 15\%, to 150 million.

\end{small}

%\acknowledgements{Many thanks to Jarno Alanko and Uwe Baier for their XBWT-construction software, and to Diego D\'iaz, Richard Durbin, Filippo Geraci, Giuseppe Italiano, Ben Langmead, Gonzalo Navarro, Pierre Peterlongo, Nicola Prezza, Giovanna Rosone, Jared Simpson, Jouni Sir\'en and Jan Studen\'y for helpful discussions.}

\inputXBWT{intro}
\inputXBWT{concepts}
\inputXBWT{contribution}
\inputXBWT{practice}

\inputXBWT{appendix}


\BiblatexSplitbibDefernumbersWarningOff

\backmatter
\printbibliography[segment=\therefsegment,heading=subbibintoc]